Fixes #241 :Added installer for talawa-api

kailash360 opened the pull request:
**What kind of change does this PR introduce?**
This feature adds an automated-installation setup for the Talawa API.

**Issue Number:**
Fixes #241 

**Did you add tests for your changes?**
No

**Snapshots/Videos:**
![image](https://user-images.githubusercontent.com/75155230/144510502-dfc2dce0-e989-4e90-90c4-2570fc4069d9.png)
![image](https://user-images.githubusercontent.com/75155230/144510540-7c5484d2-2fcf-454a-975e-60ac8ca64d8e.png)
![image](https://user-images.githubusercontent.com/75155230/144510579-98852d1e-2e6a-4e9c-bb29-b17982c97d7b.png)
![image](https://user-images.githubusercontent.com/75155230/144510865-276f9efb-b598-410b-ad31-ffeef5ab7f94.png)

**If relevant, did you update the documentation?**
Yes, I have added the **Automated Installation** section in the ```Installation.md``` file that shows how to run the installer.

**Summary**
The initial setup required several steps to go through in order to install and start the Talawa API. Now, it needs just to clone the code, and run a single command. Rest of the installation process will be taken care of by the setup module itself. More information is available at issue #241 

**Does this PR introduce a breaking change?**
No, it adds an additional feature to the code.

**Other information**
I have also added comments in the code so that the workflow can be understood easily.

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**
Yes

palisadoes reviewed (changes requested):
@kailash360 this is really good to see! Congrats! This has been needed for so long.

We are going to need some readability changes before this can be approved.
- We are going to need better comments in the code
- Please make sure the code is compatible with `pylint`, `pycodestyle`, `pep8` and `pydocstyle`,
- Make the code follow the [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html)

Here is a sample script you can use to help with the clean up of each file where you put the filename you want to check as the only argument.

```
#!/bin/bash

PYTHONPATH="${PYTHONPATH}:/colovore/projects/slurpy/lib/v2.0"
export PYTHONPATH

# List of error messages to suppress
# DISABLE_LIST='F0401,W0703,W0702'
DISABLE_LIST='W0702,W0703'

# Maximum number of local variables
MAX_LOCALS=20

pylint --reports=no --max-locals=$MAX_LOCALS --disable=$DISABLE_LIST $1 | grep -v 'using default configuration'
pycodestyle --show-source $1
pydocstyle $1
pep8 $1
```

kailash360 commented:
@palisadoes I have fixed the code using ```pylint``` and also added ```autopep8``` to auto-fix the code every time the lint is run. I have also integrated the command to run the lint script of python in  the ```lint``` command of ```package.json```

palisadoes reviewed (changes requested):
- Please make your docstrings compliant with the [Google Python Guide - Docstrings](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings)
- Please use only single quotes for string variables. You have a combination of single and double quotes.

palisadoes commented on setup/post_installation.py:
```diff
@@ -0,0 +1,20 @@
+"""Modules"""
```
The docstring does not explain what this module does. Please add more detail.
The first line of the docstring does not end in a period (full stop), Please run pydocstring to verify.

palisadoes commented on setup/lint.py:
```diff
@@ -0,0 +1,12 @@
+"""Modules"""
```
The docstring does not explain what this module does. Please add more detail.
The first line of the docstring does not end in a period (full stop), Please run pydocstring to verify.

palisadoes commented on setup/installation.py:
```diff
@@ -0,0 +1,72 @@
+"""Modules"""
```
The docstring does not explain what this module does. Please add more detail.
The first line of the docstring does not end in a period (full stop), Please run pydocstring to verify.

palisadoes commented on setup/data/About.md:
```diff
@@ -0,0 +1,11 @@
+**Talawa** is a modular open source project to manage group activities of both non-profit organizations and businesses.
```
Why is this file included?

palisadoes commented on setup/data/Input.md:
```diff
@@ -0,0 +1,3 @@
+## USER SETUP
```
Why is this file included?

palisadoes commented on setup.py:
```diff
@@ -0,0 +1,54 @@
+
+"""Display initial data"""
+utils.display_markdown('# Talawa API')
+with open('./setup/data/About.md', encoding="utf-8") as about:
```
OK, now I see the purpose of the markdown files. Please rename `data/` to `markdown/` so it is clearer.

palisadoes commented on setup.py:
```diff
@@ -0,0 +1,54 @@
+    code = compile(f.read(), VIRTUAL_ENV_PATH, 'exec')
+    exec(code, dict(__file__=VIRTUAL_ENV_PATH))
+
+"""Insttall Python packages as in requirements.txt"""
```
Typo in doc string

palisadoes commented on setup.py:
```diff
@@ -0,0 +1,54 @@
+    utils.display_markdown(user_input.read(), "white")
+installation.user_input()
+
+"""Install JavaSceipt dependencies as in package.json"""
```
Typo in docstring

palisadoes commented on setup/installation.py:
```diff
@@ -0,0 +1,72 @@
+        config.write(data)
+
+
+"""To run any command on the shell
```
This needs to be placed within the docstring of the `run` function as defined in the [Google Python Style Guide - Docstrings](https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings)

palisadoes commented on setup/lint.py:
```diff
@@ -0,0 +1,12 @@
+"""Modules"""
+import subprocess
```
We need all the python files in the repo linted. In hindsight, this should be included as a GitHub action. The Action must fail if the linting fails. 

At the moment, the code appears to only check setup.py which is insufficient. Let's discuss this after all the other fixes are done. This part of the code should be included in a separate issue.

kailash360 replied:
@palisadoes, 
I have generated a ```.pylintrc``` file using ```pylint```, and it contains the configurations of code formatting for Python files. The configs can be customized as per our need.
I have integrated the command for linting within the  ```lint``` script in ```package.json```. It would lint the entire project, and exclude some specific folders mentioned in the ```.pylintrc``` file, at ```line 23```. Upon execution, it would give the following output:
![image](https://user-images.githubusercontent.com/75155230/145634328-918abaaa-f9b8-4897-a2d7-47815241a94d.png)

I also wanted to know :
1. Whether I should create a GitHub action for the lint check, 
2. If above point is yes, then should it be another issue
3. The [```lint.py```](https://github.com/kailash360/talawa-api/blob/kailash360/setup/lint.py) file is of no use now. Should it be removed?


I have pushed the code. Do suggest changes, if any.

palisadoes commented on setup/utils.py:
```diff
@@ -0,0 +1,24 @@
+"""Modules"""
```
The docstring does not explain what this module does. Please add more detail.
The first line of the docstring does not end in a period (full stop), Please run pydocstring to verify.

palisadoes reviewed (changes requested):
Almost there! Please take some more time to really make the code readable to new contributors.

palisadoes commented:
> 

I didn't expect the linting to be done as part of the installation process. It's too late then. I meant that you should enforce these python linting rules on the various python scripts and modules.

kailash360 commented:
I have now linted the code according to the guide mentioned above.
I also ran the ```pylint``` command to verify and it gave this result.

![image](https://user-images.githubusercontent.com/75155230/144727581-5a24d023-f18c-4741-bca9-49016fca8db0.png)

I have also added a ```README.md``` inside the ```setup``` directory and explained the file structures and functionalities there.
Every function has a docstring now specifying the arguments, return values, and their types. I have also explained the flow of execution of the driver code within the ```setup.py``` file, following the conventions in the guide.

Also, the ```lint.py``` script is not a part of the installation, I have kept it within the setup module. I was not sure whether to keep it within the root folder. It runs only when the lint command is run.

Hope the code is configured well now.

palisadoes commented:
@kailash360 Sorry for the delay in the review.  Please fix the final queries I have and we should be OK.

Thanks for the hard work on this. Kudos.

palisadoes commented:
@kailash360 We need to talk about this. Check my message in slack

palisadoes commented:
Please address these outstanding items:

- We need to update the original installation.md file so that everyone can start using this code
- Does the code support windows?
- How are you ensuring that the `.env` file is placed in the right directory? Suppose someone runs the scripts from outside the Talawa-API directory, will the .env file be created in the correct location?
- Please use single quotes for all your variables. """ for docstrings.
- I see the purpose of the markdown files. Please rename the `data/` directory to `markdown/` so it is clearer.

kailash360 commented:
> Please address these outstanding items:
> 
> * We need to update the original installation.md file so that everyone can start using this code
> * Does the code support windows?
> * How are you ensuring that the `.env` file is placed in the right directory? Suppose someone runs the scripts from outside the Talawa-API directory, will the .env file be created in the correct location?
> * Please use single quotes for all your variables. """ for docstrings.
> * I see the purpose of the markdown files. Please rename the `data/` directory to `markdown/` so it is clearer.

@palisadoes, 
- I have added the [**Automated Installation**](https://github.com/kailash360/talawa-api/blob/kailash360/INSTALLATION.md#automated-installation) section inside ```INSTALLATION.md``` to explain how to setup using this code
- Yes, the code is supported for all platforms since the commands used here are of python itself, and they do not depend on a specific platform
- In my latest push, I have added one more step in the installation - *Shifting to the Talawa API directory*. It is now the first step of installation process, which ensures that the path is set to the Talawa API directory.
- For string variables, I have replaced the double quotes with single quotes, and docstrings are already enclosed within """. 
- I have now renamed the ```data``` directory to ```markdown```

It would be really helpful, if you can also clarify on the queries I have mentioned [above](https://github.com/PalisadoesFoundation/talawa-api/pull/387#discussion_r766951614)

palisadoes commented:
@kailash360 

- Whether I should create a GitHub action for the lint check, **Yes**
- If above point is yes, then should it be another issue **Yes**
- The `lint.py `file is of no use now. Should it be removed? **Yes**

Thanks for your diligence on this issue.

palisadoes commented:
@kailash360 Let me know when you are done. So we can do a final review and merge.

kailash360 commented:
@palisadoes 
I have the final changes and pushed them. Hope it passes the final review.
Also, I learnt a lot of new things in this issue alone 😁

palisadoes commented:
@kailash360 These questions in my previous review haven't been addressed.

- What should the user do if `npm` is not found? This is important as any error could cause them to choose some other software. You should have a step before this where `npm` is installed, for RedHat, Ubuntu and Windows variants.
- How will users start the application after a reboot? This just gets them running the first time. We need the command they need to run afterwards.

I know installing `npm` is trivial, but we must ensure that it's done beforehand. The installation process may change, so you may want to have a reference to the package provider's website's installation page.

Figuring out how to start the application after a reboot is also important. Please open another issue to address this. I don't want this PR to be held up much longer.

kailash360 commented:
@palisadoes 
- The first command of setup uses ```npm``` itself. In order to detect if ```npm``` is installed or not, and to install it in case it isn't installed, I was thinking about going with command files for individual OS like ```.bat``` , ```.sh``` , and each of them get executed as the first step of the setup. They will contain the script to check if ```npm``` and ```Python 3``` are installed or not, and insttall them.
 
    Alternatively, we can also create a pre-setup Python script and execute it, assuming that Python 3 is already installed. I am a bit confused as which of the above two will be a better way to deal with this situation.
   
    Please suggest if there is also some other way around.
- To restart the server, the user will simply need to run the ```npm start``` command, and the project will start. Please clarify in the new issue if you referred to something else.

palisadoes commented:
We need to think of a way to automatically restart the server if the system reboots. For example in the case of Linux, a systemctl service file. A process that requires human intervention to restart is not good enough. This should be tackled in a separate issue where the installation process also creates the service.

palisadoes commented:
Regarding the `npm` preinstallation, I think it's good enough for now to add a step in the instructions that tells people to install `npm` as a prerequisite. The step should have a link to the node website's installation page. It's tempting to do everything for the user, but node installation is best handled by the package developer, not us.

palisadoes commented:
@DangaRanga Please take a look and run to see if it works for you.

palisadoes commented:
@kailash360 In the mean time, are there any other flutter related issues you'd like to work on?

DangaRanga commented:
> @DangaRanga Please take a look and run to see if it works for you.

I apologize for the untimely review

When running the setup I had encountered the following issues: 

1. Cryptography had not installed as apart of the initial setup, which had caused the installation to fail
2. Virtualenv had not been added to PATH by default, which had caused the installation to fail

My detailed results are as follows:

### Cryptography had not installed as apart of the initial setup
![image](https://user-images.githubusercontent.com/55003601/146017129-b344a05b-8b04-4e25-9a61-01aa2e547ff4.png)

By any chance did you forget to create the virtualenv before installing the packages?

After removing my local packages it seems that the packages specified within the requirements.txt do get installed to my local site-packages directory, as opposed to a virtualenv, was this intended?

### Virtualenv had not been added to PATH by default

I had installed cryptography afterwards, however I had encountered another issue to virutualenv not being added to PATH by default. I assume that most users may also encounter a similar issue when installing for the first time, so you may have to look into creating the virtualenv without assuming that it is added to PATH by default.

![image](https://user-images.githubusercontent.com/55003601/146020862-3c550679-5af8-4f48-b49b-26b518495434.png)


One approach that I had used prior had been as follows

```
python3 -m virtualenv <path/to/virtualenv>
```

The code sample can be found here https://github.com/PalisadoesFoundation/pattoo-shared/blob/master/pattoo_shared/installation/environment.py#L19

Good job so far, I hope that this helps!

kailash360 commented:
@DangaRanga 
There was an issue with the user permissions, I have fixed it and pushed the changes.

palisadoes commented:
I got the following errors:

![image](https://user-images.githubusercontent.com/16875803/146282349-81701153-4446-4d5c-ab55-47739ba3232d.png)
![image](https://user-images.githubusercontent.com/16875803/146282431-51e26ddd-a2ee-4d0b-8655-46c69d9b15ff.png)
![image](https://user-images.githubusercontent.com/16875803/146282473-57b47ad8-4fd2-449f-8384-56545928d1cc.png)
![image](https://user-images.githubusercontent.com/16875803/146282535-0e69df2f-d869-4e75-95f3-322ec4a45664.png)
![image](https://user-images.githubusercontent.com/16875803/146282593-dc4d7382-37e7-49f2-8623-405b14e1dc44.png)
![image](https://user-images.githubusercontent.com/16875803/146282690-70bef21c-b3e4-49ab-ac68-1112ae730933.png)
![image](https://user-images.githubusercontent.com/16875803/146282762-60575541-d734-426c-b5f1-96bc6f509a4e.png)

palisadoes commented:
My `.env` file looks like this:

```
NAME=Peter Harrison
USERNAME=administrator
PASSWORD=b'gAAAAABhun0dMYaU8FTho5-xOuPhaUt_yEklks8iLhCdzhR7FNratpBxH06-0HpypMcWCmZE6usAddT3eKzpCW9i4kSNaK5O2g=='
DECRYPTION_KEY=b'kPqcyjCiSVoHtkTDFKti5bs5eAE4QgtfJAfUbOBDd-k='
ACCESS_TOKEN_SECRET=123456
REFRESH_TOKEN_SECRET=1234567890
nMONGO_DB_URL=abcdefg
```
- I purposely put in a bogus DB URL to check your error handling. 
- I think you'll need to verify whether the MongoDB URL is valid and responsive as part of the installation process.

I recognize this is a worst case scenario, but end users can do strange things.

palisadoes commented:
When I run `npm start` I get the following message which is related to the bad Mongo URI.

```
> talawa-api@1.0.0 start
> cross-env pm2-runtime start index.js --watch

2021-12-15T15:58:07: PM2 log: Launching in no daemon mode
2021-12-15T15:58:07: PM2 log: [Watch] Start watching index
2021-12-15T15:58:07: PM2 log: App [index:0] starting in -fork mode-
2021-12-15T15:58:07: PM2 log: App [index:0] online
error 2021-12-15 15:58:09 - Error while connecting to mongo database The `uri` parameter to `openUri()` must be a string, got "undefined". Make sure the first parameter to `mongoose.connect()` or `mongoose.createConnection()` is a string. {}
MongooseError: The `uri` parameter to `openUri()` must be a string, got "undefined". Make sure the first parameter to `mongoose.connect()` or `mongoose.createConnection()` is a string.
    at NativeConnection.Connection.openUri (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/connection.js:694:11)
    at /home/peter/GitHub/talawa-api/node_modules/mongoose/lib/index.js:351:10
    at /home/peter/GitHub/talawa-api/node_modules/mongoose/lib/helpers/promiseOrCallback.js:32:5
    at new Promise (<anonymous>)
    at promiseOrCallback (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/helpers/promiseOrCallback.js:31:10)
    at Mongoose._promiseOrCallback (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/index.js:1149:10)
    at Mongoose.connect (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/index.js:350:20)
    at Object.connect (/home/peter/GitHub/talawa-api/db.js:6:20)
    at serverStart (/home/peter/GitHub/talawa-api/index.js:186:20)
    at Object.<anonymous> (/home/peter/GitHub/talawa-api/index.js:204:1)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Object.<anonymous> (/home/peter/GitHub/talawa-api/node_modules/pm2/lib/ProcessContainerFork.js:33:23)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
2021-12-15T15:58:09: PM2 log: App [index:0] exited with code [1] via signal [SIGINT]
2021-12-15T15:58:09: PM2 log: App [index:0] starting in -fork mode-
2021-12-15T15:58:09: PM2 log: App [index:0] online
error 2021-12-15 15:58:10 - Error while connecting to mongo database The `uri` parameter to `openUri()` must be a string, got "undefined". Make sure the first parameter to `mongoose.connect()` or `mongoose.createConnection()` is a string. {}
MongooseError: The `uri` parameter to `openUri()` must be a string, got "undefined". Make sure the first parameter to `mongoose.connect()` or `mongoose.createConnection()` is a string.
    at NativeConnection.Connection.openUri (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/connection.js:694:11)
    at /home/peter/GitHub/talawa-api/node_modules/mongoose/lib/index.js:351:10
    at /home/peter/GitHub/talawa-api/node_modules/mongoose/lib/helpers/promiseOrCallback.js:32:5
    at new Promise (<anonymous>)
    at promiseOrCallback (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/helpers/promiseOrCallback.js:31:10)
    at Mongoose._promiseOrCallback (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/index.js:1149:10)
    at Mongoose.connect (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/index.js:350:20)
    at Object.connect (/home/peter/GitHub/talawa-api/db.js:6:20)
    at serverStart (/home/peter/GitHub/talawa-api/index.js:186:20)
    at Object.<anonymous> (/home/peter/GitHub/talawa-api/index.js:204:1)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Object.<anonymous> (/home/peter/GitHub/talawa-api/node_modules/pm2/lib/ProcessContainerFork.js:33:23)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
2021-12-15T15:58:10: PM2 log: App [index:0] exited with code [1] via signal [SIGINT]
2021-12-15T15:58:10: PM2 log: App [index:0] starting in -fork mode-
2021-12-15T15:58:10: PM2 log: App [index:0] online
error 2021-12-15 15:58:12 - Error while connecting to mongo database The `uri` parameter to `openUri()` must be a string, got "undefined". Make sure the first parameter to `mongoose.connect()` or `mongoose.createConnection()` is a string. {}
MongooseError: The `uri` parameter to `openUri()` must be a string, got "undefined". Make sure the first parameter to `mongoose.connect()` or `mongoose.createConnection()` is a string.
    at NativeConnection.Connection.openUri (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/connection.js:694:11)
    at /home/peter/GitHub/talawa-api/node_modules/mongoose/lib/index.js:351:10
    at /home/peter/GitHub/talawa-api/node_modules/mongoose/lib/helpers/promiseOrCallback.js:32:5
    at new Promise (<anonymous>)
    at promiseOrCallback (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/helpers/promiseOrCallback.js:31:10)
    at Mongoose._promiseOrCallback (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/index.js:1149:10)
    at Mongoose.connect (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/index.js:350:20)
    at Object.connect (/home/peter/GitHub/talawa-api/db.js:6:20)
    at serverStart (/home/peter/GitHub/talawa-api/index.js:186:20)
    at Object.<anonymous> (/home/peter/GitHub/talawa-api/index.js:204:1)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Object.<anonymous> (/home/peter/GitHub/talawa-api/node_modules/pm2/lib/ProcessContainerFork.js:33:23)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
2021-12-15T15:58:12: PM2 log: App [index:0] exited with code [1] via signal [SIGINT]
2021-12-15T15:58:12: PM2 log: App [index:0] starting in -fork mode-
2021-12-15T15:58:12: PM2 log: App [index:0] online
error 2021-12-15 15:58:14 - Error while connecting to mongo database The `uri` parameter to `openUri()` must be a string, got "undefined". Make sure the first parameter to `mongoose.connect()` or `mongoose.createConnection()` is a string. {}
MongooseError: The `uri` parameter to `openUri()` must be a string, got "undefined". Make sure the first parameter to `mongoose.connect()` or `mongoose.createConnection()` is a string.
    at NativeConnection.Connection.openUri (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/connection.js:694:11)
    at /home/peter/GitHub/talawa-api/node_modules/mongoose/lib/index.js:351:10
    at /home/peter/GitHub/talawa-api/node_modules/mongoose/lib/helpers/promiseOrCallback.js:32:5
    at new Promise (<anonymous>)
    at promiseOrCallback (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/helpers/promiseOrCallback.js:31:10)
    at Mongoose._promiseOrCallback (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/index.js:1149:10)
    at Mongoose.connect (/home/peter/GitHub/talawa-api/node_modules/mongoose/lib/index.js:350:20)
    at Object.connect (/home/peter/GitHub/talawa-api/db.js:6:20)
    at serverStart (/home/peter/GitHub/talawa-api/index.js:186:20)
    at Object.<anonymous> (/home/peter/GitHub/talawa-api/index.js:204:1)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Object.<anonymous> (/home/peter/GitHub/talawa-api/node_modules/pm2/lib/ProcessContainerFork.js:33:23)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
2021-12-15T15:58:14: PM2 log: App [index:0] exited with code [1] via signal [SIGINT]
2021-12-15T15:58:14: PM2 log: App [index:0] starting in -fork mode-
2021-12-15T15:58:14: PM2 log: App [index:0] online
^C2021-12-15T15:58:15: PM2 log: [Watch] Stop watching index
2021-12-15T15:58:15: PM2 log: Stopping app:index id:0
2021-12-15T15:58:15: PM2 log: Stopping app:index id:0
2021-12-15T15:58:15: PM2 log: App [index:0] exited with code [0] via signal [SIGINT]
2021-12-15T15:58:15: PM2 log: pid=3048879 msg=process killed
2021-12-15T15:58:15: PM2 log: pid=3048879 msg=process killed
2021-12-15T15:58:15: PM2 log: PM2 successfully stopped
2021-12-15T15:58:15: PM2 log: PM2 successfully stopped
```

palisadoes commented:
- You'll need to specify installing `MongoDB` as part of the automated installation
- You'll need to do some basic `MongoDB` checking as part of the installation.
- I think we'll need to rename the `Standard Installation` section and call it `Expert Installation` with a Docker subsection 
- This new method should be called the `Express Installation` section in the installation guide as it makes things easy and only asks for a few variables. 
- You'll need to find a way to only update the required sections of an `.env` file if it already exists. Overwriting an existing `.env` file could be disastrous.
- Would it make sense for the installer to generate the `ACCESS_TOKEN_SECRET` and `REFRESH_TOKEN_SECRET` for simplicity or does this need to come from MongoDB?
- The start() function in `setup/post_installation.py` doesn't capture the STDERR and STDOUT together. If there is STERR, it assumes there is no STDOUT and there clearly is some output that is not being shown.

I know we are asking a lot, but the installation needs to be bulletproof. It's the API's greatest weakness at the moment.

palisadoes commented:
@kailash360 I think the easiest way around this issue is not to run `npm start` as part of the installation. Just let the user know that they should run the command afterwards. But we should at least verify that the MongoDB URL is OK as well as the remaining `.env` parameters. What do you think?

kailash360 commented:
Yes, we can check if the given URI matches with a MongoDB URI or not. 

Also, it would be helpful if you can clarify on what needs to be done for the ```Access token``` and ```Refresh token```.

palisadoes commented:
> Yes, we can check if the given URI matches with a MongoDB URI or not.
> 
> Also, it would be helpful if you can clarify on what needs to be done for the `Access token` and `Refresh token`.

- @sumitra19jha @cyberwake @DMills27 what is the `Access token` and `Refresh token` used for in the `.env` file, and what format should it have?
- @kailash360 Please see how you can verify the MongoDB connectivity as part of the installation. Don't do the `npm start`. Make the user do that, because we don't want them to think they have to re-install every time they want to restart the app. I know it's unlikely, but we need to simplify things. This strategy shows that in all installation cases they have to run `npm start`

kailash360 commented:
@palisadoes 
A major issue that I observed in the scripts is that the virtual environment is not getting activated by the scripts. I tried running both ```virtualenv venv``` and ```python -m venv ./venv``` commands in the script, but they are only creating the virtual environment, and not activating it. Can you suggest any solution to this?

sandeep-v1404 commented:
> @palisadoes A major issue that I observed in the scripts is that the virtual environment is not getting activated by the scripts. I tried running both `virtualenv venv` and `python -m venv ./venv` commands in the script, but they are only creating the virtual environment, and not activating it. Can you suggest any solution to this?

You can do it by running `source venv/bin/activate`

kailash360 commented:
> > @palisadoes A major issue that I observed in the scripts is that the virtual environment is not getting activated by the scripts. I tried running both `virtualenv venv` and `python -m venv ./venv` commands in the script, but they are only creating the virtual environment, and not activating it. Can you suggest any solution to this?
> 
> This command `python -m venv ./venv` creates a virtual environment in the root folder. To activate it, you can run
> 
> * `source venv/bin/activate` for Linux.
> * `venv\Scripts\activate` for Windows.

Sorry for the mistake above,
I had run these commanda actually. And they weren't executed as expected.

sandeep-v1404 commented:
> > > @palisadoes A major issue that I observed in the scripts is that the virtual environment is not getting activated by the scripts. I tried running both `virtualenv venv` and `python -m venv ./venv` commands in the script, but they are only creating the virtual environment, and not activating it. Can you suggest any solution to this?
> > 
> > 
> > This command `python -m venv ./venv` creates a virtual environment in the root folder. To activate it, you can run
> > 
> > * `source venv/bin/activate` for Linux.
> > * `venv\Scripts\activate` for Windows.
> 
> Sorry for the mistake above, I had run these commanda actually. And they weren't executed as expected.

Not a problem. Is the script able to create a virtual environment?

kailash360 commented:
> > > > @palisadoes A major issue that I observed in the scripts is that the virtual environment is not getting activated by the scripts. I tried running both `virtualenv venv` and `python -m venv ./venv` commands in the script, but they are only creating the virtual environment, and not activating it. Can you suggest any solution to this?
> > > 
> > > 
> > > This command `python -m venv ./venv` creates a virtual environment in the root folder. To activate it, you can run
> > > 
> > > * `source venv/bin/activate` for Linux.
> > > * `venv\Scripts\activate` for Windows.
> > 
> > 
> > Sorry for the mistake above, I had run these commanda actually. And they weren't executed as expected.
> 
> Not a problem. Is the script able to create a virtual environment?

Yes, after executing the ```virtualenv venv``` command via the script, the ```venv``` folder for virtual environment was created. However it was not activated after running the activation commands.

sandeep-v1404 commented:
> > > > > @palisadoes A major issue that I observed in the scripts is that the virtual environment is not getting activated by the scripts. I tried running both `virtualenv venv` and `python -m venv ./venv` commands in the script, but they are only creating the virtual environment, and not activating it. Can you suggest any solution to this?
> > > > 
> > > > 
> > > > This command `python -m venv ./venv` creates a virtual environment in the root folder. To activate it, you can run
> > > > 
> > > > * `source venv/bin/activate` for Linux.
> > > > * `venv\Scripts\activate` for Windows.
> > > 
> > > 
> > > Sorry for the mistake above, I had run these commanda actually. And they weren't executed as expected.
> > 
> > 
> > Not a problem. Is the script able to create a virtual environment?
> 
> Yes, after executing the `virtualenv venv` command via the script, the `venv` folder for virtual environment was created. However it was not activated after running the activation commands.

Okay. Check with this command once `venv/Scripts/Activate`.

kailash360 commented:
> > > > > > @palisadoes A major issue that I observed in the scripts is that the virtual environment is not getting activated by the scripts. I tried running both `virtualenv venv` and `python -m venv ./venv` commands in the script, but they are only creating the virtual environment, and not activating it. Can you suggest any solution to this?
> > > > > 
> > > > > 
> > > > > This command `python -m venv ./venv` creates a virtual environment in the root folder. To activate it, you can run
> > > > > 
> > > > > * `source venv/bin/activate` for Linux.
> > > > > * `venv\Scripts\activate` for Windows.
> > > > 
> > > > 
> > > > Sorry for the mistake above, I had run these commanda actually. And they weren't executed as expected.
> > > 
> > > 
> > > Not a problem. Is the script able to create a virtual environment?
> > 
> > 
> > Yes, after executing the `virtualenv venv` command via the script, the `venv` folder for virtual environment was created. However it was not activated after running the activation commands.
> 
> Okay. Check with this command once `venv/Scripts/Activate`.

Nope, didn't work

sumitra19jha commented:
![Screenshot 2021-12-18 at 8 05 19 PM](https://user-images.githubusercontent.com/35868598/146644759-7130b775-2bf2-480a-9be7-e22aa86a4658.png)

Please look here

sumitra19jha commented on setup.py:
```diff
@@ -0,0 +1,129 @@
+installation.user_input()
+
+# 7. Install JavaScript dependencies
+utils.display_markdown("# INSTALLING JAVASCRIPT DEPENDENCIES", "white")
```
Remove Javascript and use Project

kailash360 replied:
Done

sumitra19jha commented on setup/README.md:
```diff
@@ -0,0 +1,36 @@
+
+A brief explanation of each of the contents of this file is given below :
+
+- installation.py : This file contails all the functions that are necessary to execute the installation of the Talawa API. 
```
Can you please add a brief description of each function especially user_input which is responsible for env creation

kailash360 replied:
Description for each and every function is present within the function body, following the [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html). you can read more about the ```user_input``` function from [```installation.py```](https://github.com/kailash360/talawa-api/blob/kailash360/setup/installation.py#L37)

sumitra19jha replied:
You are following the google python style guide in the python file. I am asking to change documentation (.md). Please add a function description here too (brief one).

kailash360 replied:
I have added the function descriptions now.

kailash360 commented:
@palisadoes 
I have now addressed the issues mentioned [here](https://github.com/PalisadoesFoundation/talawa-api/pull/387#issuecomment-995319763). Also, two different starting scripts are executed based on the operating system of the user.

sumitra19jha commented on INSTALLATION.md:
````diff
@@ -91,6 +91,27 @@ Follow these steps to get the api running using npm
+   ```sh
+   npm run setup
+   ```
 
````
Can you please add a small note here and put the reference of `setup/README.md` like `For more information on automated installation visit [here](url)`

kailash360 replied:
Done

sumitra19jha commented on setup/README.md:
```diff
@@ -0,0 +1,36 @@
+- markdown : This subdirectory contains the data displayed during installation of the Talawa API
+    - About.md - This is the information that is displayed when the installation of the Talawa API begins 
+    - Input.md - This file contains the instructions that are shown when the user provides the program credentials as input. 
+
```
Please delete `__init__.py` file

kailash360 replied:
Done

sumitra19jha reviewed (changes requested):
Great work in this PR Kailash

sumitra19jha commented:
> > Yes, we can check if the given URI matches with a MongoDB URI or not.
> > Also, it would be helpful if you can clarify on what needs to be done for the `Access token` and `Refresh token`.
> 
> * @sumitra19jha @CyberWake @DMills27 what is the `Access token` and `Refresh token` used for in the `.env` file, and what format should it have?
> * @kailash360 Please see how you can verify the MongoDB connectivity as part of the installation. Don't do the `npm start`. Make the user do that, because we don't want them to think they have to re-install every time they want to restart the app. I know it's unlikely, but we need to simplify things. This strategy shows that in all installation cases they have to run `npm start`

They are secret for creating and verifying token

sumitra19jha commented:
> ![Screenshot 2021-12-18 at 8 05 19 PM](https://user-images.githubusercontent.com/35868598/146644759-7130b775-2bf2-480a-9be7-e22aa86a4658.png)
> 
> Please look here

@kailash360  Please look into this issue as well

kailash360 commented on setup/installation.py:
```diff
@@ -0,0 +1,125 @@
+    Returns:
+        None
+    """
+    name = input("Name: ")
```
This input field is used to take the name of the super administrator. It is finally stored in the environment variable.

sumitra19jha replied:
My question why are we using this input in the env. What special purpose does it serve (like MongoDB uri or access token which is very important) why you are creating this?

sumitra19jha replied:
Previous env file doesn't have this

kailash360 replied:
It was actually mentioned in the [Issue #241](https://github.com/PalisadoesFoundation/talawa-api/issues/241) itself under the **Installer Operations** section. So, I proceeded to add them. 
If you say, I can remove them. It won't matter much.

sumitra19jha replied:
To the best of my knowledge, the unused variables should be omitted. We can add them later if needed. @palisadoes Please advise what to do?

kailash360 replied:
@sumitra19jha 
Should I also remove the password field?

kailash360 commented on setup/installation.py:
```diff
@@ -0,0 +1,125 @@
+        None
+    """
+    name = input("Name: ")
+    username = input("Username: ")
```
This input field like the name, takes input of the username from the super administrator and saves it as an evironment variable.

kailash360 commented on setup/installation.py:
```diff
@@ -0,0 +1,125 @@
+    """
+    name = input("Name: ")
+    username = input("Username: ")
+    password = getpass.getpass("Password: ")
```
This input field like the name, takes input of the password from the super administrator and saves it as an environment variable.

kailash360 commented:
@sumitra19jha 
I have made the changes as asked.

palisadoes commented:
- @sumitra19jha 
- @DangaRanga 

Please review. This is a really important PR for us to merge. Try to install it on your systems and see if there are any problems that could make the UX undesirable.

palisadoes commented:
@kailash360 You have some conflicting files. Please update the PR.

kailash360 commented:
> @kailash360 You have some conflicting files. Please update the PR.

@palisadoes 
The conflicts have now been resolved

kailash360 commented:
> * @sumitra19jha
> * @DangaRanga
> 
> Please review. This is a really important PR for us to merge. Try to install it on your systems and see if there are any problems that could make the UX undesirable.

The instructions in the ```INSTALLATION.md``` has been updated. Kindly go through them once, before starting with the installation process.

palisadoes commented:
@kailash360 Linting is failing

kailash360 commented:
> @kailash360 Linting is failing

Fixed the lint now

palisadoes commented:
Sorry for the delay. This is why: 

We had an extensive discussion for `talawa-admin` about creating an installer which was addressed in this PR.

- https://github.com/PalisadoesFoundation/talawa-admin/pull/208

The thought was that installing python solely for purposes of installation was inappropriate and that installing using node, which is a prerequisite for `talawa-admin` operation, would be a better choice.

@kailash360, @yasharth291 do you think this is a valid approach?

kailash360 commented:
> Sorry for the delay. This is why:
> 
> We had an extensive discussion for `talawa-admin` about creating an installer which was addressed in this PR.
> 
> * [Talawa admin installer in JavaScript talawa-admin#208](https://github.com/PalisadoesFoundation/talawa-admin/pull/208)
> 
> The thought was that installing python solely for purposes of installation was inappropriate and that installing using node, which is a prerequisite for `talawa-admin` operation, would be a better choice.
> 
> @kailash360, @yasharth291 do you think this is a valid approach?

Yes, 
Considering the codebase of this application, and ease of accessibility, NodeJS is surely a better choice. I too was thinking of using JavaScript as the tool. However Python was already mentioned in the issue, so I proceeded with the latter.

palisadoes commented:
I suggested python as I wasn't aware of the scripting capabilities of `node.js`. 
How much effort would be required to change the installer to use javascript?

kailash360 commented:
> I suggested python as I wasn't aware of the scripting capabilities of `node.js`. How much effort would be required to change the installer to use javascript?

This will involve shifting the entire codebase to JavaScript. Since the concepts and flow remains the same, it will not take too much of an effort.

palisadoes commented:
> > I suggested python as I wasn't aware of the scripting capabilities of `node.js`. How much effort would be required to change the installer to use javascript?
> 
> @palisadoes This will involve shifting the entire codebase to JavaScript. Since the concepts and flow remains the same, it will not take too much of an effort. Should I shift it to Node?

You mean just the python parts you did, not the entire API codebase correct?

kailash360 commented:
> > > I suggested python as I wasn't aware of the scripting capabilities of `node.js`. How much effort would be required to change the installer to use javascript?
> > 
> > 
> > @palisadoes This will involve shifting the entire codebase to JavaScript. Since the concepts and flow remains the same, it will not take too much of an effort. Should I shift it to Node?
> 
> You mean just the python parts you did, not the entire API codebase correct?

Yes, just the ```.py``` files that I have created

palisadoes commented:
Please proceed.

kailash360 commented:
@palisadoes 
Can I close this PR and create a new one with the JavaScript code? This PR has been stuck for too long, and contains a long discussion.

palisadoes commented:
Yes. I'll close it.
